apis/Google.Storage.V1/Google.Storage.V1/SignedUrlUtils.cs

```diff
@@ -0,0 +1,410 @@
+﻿// Copyright 2015 Google Inc. All Rights Reserved.
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Api.Gax;
+using Google.Apis.Auth.OAuth2;
+using Google.Apis.Json;
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Security.Cryptography;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Google.Storage.V1
+{
+    /// <summary>
+    /// Class which helps create signed URLs which can be used to provide limited access to specific buckets and objects
+    /// to anyone in possession of the URL, regardless of whether they have a Google account.
+    /// </summary>
+    /// <remarks>
+    /// See https://cloud.google.com/storage/docs/access-control/signed-urls for more information on signed URLs.
+    /// </remarks>
+    public static class SignedUrlUtils
+    {
+        private const string GoogHeaderPrefix = "x-goog-";
+        private const string GoogEncryptionKeyHeader = "x-goog-encryption-key";
+        private const string GoogEncryptionKeySHA256Header = "x-goog-encryption-key-sha256";
+        private const string StorageHost = "https://storage.googleapis.com";
+        private static readonly DateTimeOffset UnixEpoch = new DateTimeOffset(new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc), TimeSpan.Zero);
+
+        /// <summary>
+        /// Creates a signed URL which can be used to provide limited access to specific buckets and objects to anyone
+        /// in possession of the URL, regardless of whether they have a Google account.
+        /// </summary>
+        /// <param name="bucket">The name of the bucket. Must not be null.</param>
+        /// <param name="objectName">The name of the object within the bucket. May be null, in which case the signed URL
+        /// will refer to the bucket instead of an object.</param>
+        /// <param name="credentialFilePath">The path to the JSON key file for a service account. Must not be null.</param>
+        /// <param name="duration">The length of time for which the signed URL should remain usable.</param>
+        /// <param name="request">A sample request for which the signed URL might be used. May be null.</param>
+        /// <remarks>
```

Comment:
I'd normally put these just below the summary, above param. Happy to alter later though - especially as we'll be modifying things to change the create/use lifecycle.